.circleci/config.yml

```diff
@@ -1,8 +1,408 @@
+          paths:
+            - tmp/go-cache
+      - persist_to_workspace:
+          root: << parameters.circleci_home >>
+          paths:
+            - project
+            - go
+            - gimme
+            - .gimme
+
+  cache_libsodium:
+    description: Cache libsodium for build
+    steps:
+      - run:
+          name: Get libsodium md5
+          command: |
+            mkdir -p tmp
+            find crypto/libsodium-fork -type f -exec openssl md5 "{}" + > tmp/libsodium.md5
+      - save_cache:
+          key: 'libsodium-fork-{{ .Environment.CIRCLE_STAGE }}-{{ checksum "tmp/libsodium.md5" }}'
+          paths:
+            - crypto/libs
+
+  restore_libsodium:
+    description: Restore libsodium for build
+    steps:
+      - run:
+          name: Get libsodium md5
+          command: |
+            mkdir -p tmp
+            find crypto/libsodium-fork -type f -exec openssl md5 "{}" + > tmp/libsodium.md5
+      - restore_cache:
+          keys:
+            - 'libsodium-fork-{{ .Environment.CIRCLE_STAGE }}-{{ checksum "tmp/libsodium.md5" }}'
+
+  generic_shorttest:
+    description: Run short tests from build workspace, for re-use by diferent architectures
+    parameters:
+      circleci_home:
+        type: string
+        default: "/home/circleci"
+      result_subdir:
+        type: string
+    steps:
+      - attach_workspace:
+          at: << parameters.circleci_home >>
+      - run: mkdir -p /tmp/results/<< parameters.result_subdir >>
+      - restore_cache:
+          keys:
+            - 'go-cache-{{ .Environment.CIRCLE_STAGE }}-'
+      - run:
+          name: Run short tests
+          no_output_timeout: 30m
+          command: |
+            set -e
+            set -x
+            export PATH=$(echo "$PATH" | sed -e 's|:<< parameters.circleci_home >>/\.go_workspace/bin||g' | sed -e 's|:/usr/local/go/bin||g')
+            export KMD_NOUSB=True
+            export GOPATH="<< parameters.circleci_home >>/go"
+            export PATH="${PATH}:${GOPATH}/bin"
+            export ALGORAND_DEADLOCK=enable
+            GOLANG_VERSION=$(./scripts/get_golang_version.sh)
+            eval "$(~/gimme "${GOLANG_VERSION}")"
+            scripts/configure_dev.sh
+            scripts/buildtools/install_buildtools.sh -o "gotest.tools/gotestsum"
+            PACKAGES="$(go list ./... | grep -v /go-algorand/test/)"
+            export PACKAGE_NAMES=$(echo $PACKAGES | tr -d '\n')
+            export PARTITION_TOTAL=$CIRCLE_NODE_TOTAL
+            export PARTITION_ID=$CIRCLE_NODE_INDEX
+            export PARALLEL_FLAG="-p 1"
+            gotestsum --format pkgname --junitfile /tmp/results/<< parameters.result_subdir >>/results.xml --jsonfile /tmp/results/<< parameters.result_subdir >>/testresults.json -- --tags "sqlite_unlock_notify sqlite_omit_load_extension" -short -race -timeout 1h -coverprofile=coverage.txt -covermode=atomic -p 1 $PACKAGE_NAMES
```

Comment:
I think these scripts should be rolled into the CI scripts at `scripts/travis`

There are subtle differences between what you have here and the scripts.

Also, there is a lot of code duplication, for example in the current scripts gotestsum is declared once in the Makefile, now it's installed and called in many places.

Reply:
It seems hard to maintain scripts that handle both Travis and Circle runs, especially since we are basically trying to cut up what Travis does. Re: code duplication, that is a matter of personal preference, but I will point out that if later we remove Travis automation, there will be less duplication, and a lot of this can migrate to CircleCI-specific scripts. However IMO total DRY-proofing this PR seems like a big request given the intermediate state we want to be in running Travis and Circle side-by-side.

Reply:
I disagree, these scripts are already platform agnostic. We've been gradually removing the travis specific aspects for a long time. Most of the heavy lifting happens in the Makefile anyway.

For example, I think this section could be replaced with:
```
            export KMD_NOUSB=True
            export PARTITION_TOTAL=$CIRCLE_NODE_TOTAL
            export PARTITION_ID=$CIRCLE_NODE_INDEX
            scripts/travis/test.sh
```

To take it one more step you could set those 3 environment variables in `test.sh`, I don't think that part is as important.

Reply:
Moving all the environment things into "if CIRCLECI" sections is definitely something you could do to bring the remaining stuff into the existing scripts.. I haven't tested whether Makefile dependency resolution works with the kind of workspace-passing between jobs/VMs that Circle and Github Actions provide though, because file timestamps would change, that seems potentially tricky.